eboutic/forms.py

```diff
@@ -0,0 +1,168 @@
+    implicitly called.
+    """
+
+    # check the json is an array containing non-nested objects.
+    # values must be strings or numbers
+    # this is matched :
+    # [{"id": 4, "name": "[PROMO 22] badges", "unit_price": 2.3, "quantity": 2}]
+    # and this is not :
+    # [{"id": {"nested_id": 10}, "name": "[PROMO 22] badges", "unit_price": 2.3, "quantity": 2}]
+    # and neither does this :
+    # [{"id": ["nested_id": 10], "name": "[PROMO 22] badges", "unit_price": 2.3, "quantity": 2}]
+    json_cookie_re = re.compile(
+        r"^\[\s*(\{\s*(\"[^\"]*\":\s*(\"[^\"]{0,64}\"|\d{0,5}\.?\d+),?\s*)*\},?\s*)*\s*\]$"
+    )
+
+    def __init__(self, request: HttpRequest):
+        self.user = request.user
+        self.cookies = request.COOKIES
+        self.error_messages = set()
+        self.correct_cookie = []
+
+    def clean(self) -> None:
+        """
+        Perform all the check, but return nothing.
+        To know if the form is valid, the `is_valid()` method must be used.
+
+        The form shall be considered as valid if it meets all the following conditions :
+            - it contains a "basket_items" key in the cookies of the request given in the constructor
+            - this cookie is a list of objects formatted this way : `[{'id': <int>, 'quantity': <int>,
+             'name': <str>, 'unit_price': <float>}, ...]`. The order of the fields in each object does not matter
+            - all the ids are positive integers
+            - all the ids refer to products available in the EBOUTIC
+            - all the ids refer to products the user is allowed to buy
+            - all the quantities are positive integers
+        """
+        basket = self.cookies.get("basket_items", None)
+        if basket is None or basket in ("[]", ""):
+            self.error_messages.add("You have no basket")
+            return
+        # check that the json is not nested before parsing it to make sure
+        # malicious user can't ddos the server with deeply nested json
+        if not BasketForm.json_cookie_re.match(basket):
+            self.error_messages.add("The request was badly formatted.")
+            return
+        try:
+            basket = json.loads(basket)
+        except json.JSONDecodeError:
+            self.error_messages.add("The request was badly formatted.")
+            return
+        if type(basket) is not list or len(basket) == 0:
+            self.error_messages.add("The request was badly formatted.")
+            return
+        eboutique = Counter.objects.get(type="EBOUTIC")
+        user_is_subscribed = self.user.subscriptions.exists()
+        for item in basket:
+            expected_keys = {"id", "quantity", "name", "unit_price"}
+            if type(item) is not dict or set(item.keys()) != expected_keys:
+                self.error_messages.add("One or more items are badly formatted.")
+                continue
+            if type(item["id"]) is not int or item["id"] < 0:
```

Comment:
```suggestion
            if not isinstance(item["id"], int) or item["id"] < 0:
```
À appliquer partout où ce genre de chose est fait, c'est plus idiomatique et plus clair (selon moi) à lire

Reply:
Ca ne me choque pas. Personnelement, je préfère `type(X) is Y`